### PR TITLE
#111 Fix issue where scenarios could miss days and add corresponding unit test

### PR DESCRIPTION
**Pull request by ofrancon**
Issue is that we assumed each country has the same number of days in the Oxford dataset. This is not true, especially when brand new data is added for some countries but not for others (not available yet), and hence dates must be handled by country/region.

### CODE REVIEW

File: validation/scenario_generator.py [hunk 1]
```diff
@@ -136,14 +136,18 @@ def generate_scenario(start_date_str, end_date_str, raw_df, countries=None, scen
...
-        # Append the new rows
-        ips_df = ips_df.append(future_rows_df)
-        # Sort
-        ips_df.sort_values(by=ID_COLS, inplace=True)
+            # Add the new rows
+            if new_rows:
+                new_rows_df = pd.DataFrame(new_rows, columns=ips_df.columns)
+                # Delete any old row that might have been replaced by a scenario one for this country / region
+                replaced_dates = list(new_rows_df["Date"].unique())
+                rows_to_drop = ips_df[(ips_df.CountryName == country) &
```
Comment (ofrancon) [inline validation/scenario_generator.py:144]: We were blindly removing dates for which we had generated a scenario, assuming they'd be replaced by the scenario ones. But scenario dates can be different by country, when one country has 1 more day of data than other for instance. So this logic is now done at the country / region level.

File: validation/tests/test_scenario_generator.py [hunk 2]
```diff
@@ -400,9 +401,36 @@ def test_generate_scenario_mind_the_gap_freeze_all_countries(self):
...
-                         f"Not the expected number of rows in the generated scenario:"
-                         f" {nb_unique_geo} geos times {nb_days_since_inception} days")
-
+        # For each country, assert the scenario contains the expected number of days
+        for country in self.latest_df.CountryName.unique():
+            all_regions = self.latest_df[self.latest_df.CountryName == country].RegionName.unique()
+            for region in all_regions:
+                ips_gdf = scenario_df[(scenario_df.CountryName == country) &
+                                      (scenario_df.RegionName == region)]
+                self.assertEqual(nb_days_since_inception, len(ips_gdf), f"Not the expected number of days"
```
Comment (ofrancon) [inline validation/tests/test_scenario_generator.py:412]: Made it easier to spot which country has a bad scenario

File: validation/tests/test_scenario_generator.py [hunk 3]
```diff
@@ -400,9 +402,36 @@ def test_generate_scenario_mind_the_gap_freeze_all_countries(self):
...
+                                                                        f" for {country} / {region}")
+
+    def test_generate_scenario_mind_the_gap_freeze_dates_mismatch(self):
```
Comment (ofrancon) [inline validation/tests/test_scenario_generator.py:415]: Now testing for this particular case

**Review by dsargent — APPROVED** [review]
lgtm

**Pull request merged**